decode.go

```diff
@@ -57,6 +58,22 @@ func newParser(b []byte) *parser {
 	return &p
 }
 
+func newFileParser(r io.Reader) *parser {
+	p := parser{}
+	if !yaml_parser_initialize(&p.parser) {
+		panic("failed to initialize YAML emitter")
+	}
+
+	yaml_parser_set_input_file(&p.parser, r)
+
+	p.skip()
```

Comment:
I'm not sure we should be calling skip here -  it means that NewDecoder can fail, and I don't think it's a great idea to make that return an error (it's nice if the YAML API mirrors the encoding/json API).